model/task_queue.go

```diff
@@ -74,7 +79,57 @@ func (self *TaskQueue) Length() int {
 }
 
 func (self *TaskQueue) NextTask() *TaskQueueItem {
-	return &self.Queue[0]
+	for _, it := range self.Queue {
+		// Always return a task if the task group is empty.
+		if it.Group == "" {
+			return &it
+		}
+		// Otherwise, return the task if it is running on fewer than its task group's max hosts.
+		spec := TaskSpec{
+			Group:         it.Group,
+			BuildVariant:  it.BuildVariant,
+			ProjectID:     it.Project,
+			Version:       it.Version,
+			GroupMaxHosts: it.GroupMaxHosts,
+		}
+		if shouldRun := shouldRunTaskGroup(it.Id, spec); shouldRun {
+			return &it
+		}
+	}
+	return nil
+}
+
+// shouldRunTaskGroup returns true if the number of hosts running a task is less than the maximum for that task group.
+func shouldRunTaskGroup(taskId string, spec TaskSpec) bool {
+	// Get number of hosts running this spec.
+	hosts, err := host.Find(host.ByTaskSpec(spec.Group, spec.BuildVariant, spec.ProjectID, spec.Version))
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":    "error finding hosts for spec",
+			"task_id":    taskId,
+			"queue_item": spec,
+		}))
+		return false
+	}
+	// If the group is running on 0 hosts, return true early.
+	if len(hosts) == 0 {
+		return true
+	}
+	t, err := task.FindOneId(taskId)
+	if err != nil || t == nil {
+		grip.Error(message.Fields{
+			"error":      errors.WithStack(err),
+			"message":    "error finding task for spec",
+			"task_id":    taskId,
+			"queue_item": spec,
+		})
+		return false
+	}
+	// If this spec is running on fewer hosts than max_hosts, dispatch this task.
+	if len(hosts) < spec.GroupMaxHosts {
+		return true
+	}
+	return false
```

Comment:
can we have methods in the host package that take a TaskSpec as an argument and return current size of task group? rather than putting this logic here, which seems wrong.

Reply:
not without an important cycle, but I did move more logic into the hosts package